src/wallet/wallet.cpp

```diff
@@ -431,22 +446,240 @@ ZcashdUnifiedSpendingKey CWallet::GenerateNewUnifiedSpendingKey() {
+        }
+
+        // Find a working diversifier and construct the associated address.
+        auto foundAddress = ufvk.FindAddress(j, receiverTypes);
+        auto diversifierIndex = foundAddress.second;
+
+        // Persist the newly created address to the keystore
+        AddUnifiedAddress(ufvkid, foundAddress);
+
+        if (hasTransparent) {
+            // Regenerate the secret key for the transparent address and add it to
+            // the wallet.
+            auto seed = GetMnemonicSeed().value();
+            auto b44 = libzcash::Bip44AccountChains::ForAccount(seed, BIP44CoinType(), accountId).value();
+            auto key = b44.DeriveExternal(j.ToTransparentChildIndex().value()).value();
+            AddTransparentSecretKey(seed.Fingerprint(), key, ufvkid);
+        }
+
+        // Save the metadata for the generated address so that we can re-derive
+        // it in the future.
+        ZcashdUnifiedAddressMetadata addrmeta(ufvkid, foundAddress.second, receiverTypes);
+
+        // we can safely ignore the return value here; we know that we're adding a new
+        // set of receivers given the receivers.has_value() check above.
+        mapUnifiedFullViewingKeyMeta[ufvkid].SetReceivers(diversifierIndex, receiverTypes);
+        if (fFileBacked) {
+            CWalletDB(strWalletFile).WriteUnifiedAddressMetadata(addrmeta);
+        }
+        return std::make_pair(foundAddress.first, addrmeta);
+    } else {
+        return AddressGenerationError::NoSuchAccount;
+    }
+}
+
+bool CWallet::LoadUnifiedFullViewingKey(const libzcash::UnifiedFullViewingKey &key)
+{
+    auto ufvkid = key.GetKeyID(Params());
+    auto zufvk = ZcashdUnifiedFullViewingKey::FromUnifiedFullViewingKey(key);
+    if (mapUnifiedFullViewingKeyMeta.count(ufvkid) > 0) {
+        // restore unified addresses that have been previously generated to the
+        // keystore
+        for (const auto &[j, receiverTypes] : mapUnifiedFullViewingKeyMeta.at(ufvkid).GetAllReceivers()) {
+            auto addr = zufvk.Address(j, receiverTypes).value();
+            AddUnifiedAddress(ufvkid, std::make_pair(addr, j));
+        }
+    }
+    return CCryptoKeyStore::AddUnifiedFullViewingKey(ufvkid, zufvk);
+}
+
+bool CWallet::LoadUnifiedKeyMetadata(const ZcashdUnifiedSpendingKeyMetadata &skmeta)
+{
+    AssertLockHeld(cs_wallet); // mapUnifiedSpendingKeyMeta
+    auto metaKey = std::make_pair(skmeta.GetSeedFingerprint(), skmeta.GetAccountId());
+    mapUnifiedSpendingKeyMeta.insert({metaKey, skmeta});
+    if (mapUnifiedFullViewingKeyMeta.count(skmeta.GetKeyID()) == 0) {
+        return mapUnifiedFullViewingKeyMeta[skmeta.GetKeyID()].SetAccountId(skmeta.GetAccountId());
+    }
+
+    return true;
+}
+
+bool CWallet::LoadUnifiedAddressMetadata(const ZcashdUnifiedAddressMetadata &addrmeta)
+{
+    AssertLockHeld(cs_wallet); // mapUnifiedSpendingKeyMeta
+    auto ufvk = GetUnifiedFullViewingKey(addrmeta.GetKeyID());
+    if (ufvk.has_value()) {
```

Comment:
Hmm; is there a wallet load race condition here? This requires that `LoadUnifiedFullViewingKey` is always called before the corresponding `LoadUnifiedAddressMetadata`, which seems entirely dependent on internal details about how BDB stores its key-value pairs, and how it reads them out.

It feels like we either need to always be able to load `ZcashdUnifiedAddressMetadata`, or we need to have a single metadata object per UFVK (instead of separate metadata objects for the USK, UFVK, and UAs), that we update and re-serialize as necessary.

Reply:
Both `LoadUnifiedAddressMetadata` and `LoadUnifiedFullViewingKey` behave in a fashion such that if one doesn't restore the address to the keystore, the other will. I should definitely add some comments to point out this interaction though, because it has the potential to be a bit fragile.